README.md


<div id="badges"><!-- pkgdown markup -->
<a href="https://CRAN.R-project.org/web/checks/check_results_TopDom.html"><img border="0" src="https://www.r-pkg.org/badges/version/TopDom" alt="CRAN check status"/></a> <a href="https://github.com/HenrikBengtsson/TopDom/actions?query=workflow%3AR-CMD-check"><img border="0" src="https://github.com/HenrikBengtsson/TopDom/actions/workflows/R-CMD-check.yaml/badge.svg?branch=develop" alt="R CMD check status"/></a>     <a href="https://app.codecov.io/gh/HenrikBengtsson/TopDom"><img border="0" src="https://codecov.io/gh/HenrikBengtsson/TopDom/branch/develop/graph/badge.svg" alt="Coverage Status"/></a> 
</div>

# TopDom: An Efficient and Deterministic Method for Identifying Topological Domains in Genomes 

The TopDom method identifies topological domains in genomes from Hi-C sequence data (Shin et al., 2016).  The authors published an implementation of their method as an R script (two different versions; also available in this package).  This package originates from those original TopDom R scripts and provides help pages adopted from the original TopDom PDF documentation.  It also provides a small number of bug fixes to the original code.


## Citing TopDom - the method and the package

Whenever using the TopDom method, please cite:

* Hanjun Shin, Yi Shi, Chao Dai, Harianto Tjong, Ke Gong, Frank Alber, Xianghong Jasmine Zhou; TopDom: an efficient and deterministic method for identifying topological domains in genomes, Nucleic Acids Research, Volume 44, Issue 7, 20 April 2016, Pages e70, [10.1093/nar/gkv1505](https://doi.org/10.1093/nar/gkv1505). PMCID: [PMC4838359](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4838359/), PMID: 26704975.

Whenever using the **TopDom** package, please cite:

* Henrik Bengtsson, Hanjun Shin, Harris Lazaris, Gangqing Hu and Xianghong Zhou (2020). R Package TopDom: An Efficient and Deterministic Method for Identifying Topological Domains in Genomes. R package version 0.10.1. <https://CRAN.R-project.org/package=TopDom>

The above information is also available as plain text as well as BibTeX entries via `citation("TopDom")`.


## Background

Please note that I, Henrik Bengtsson, is _not_ the developer of the TopDom method (Shin et al. 2016), or the (original) TopDom code available from the TopDom authors.   As I needed the TopDom method in a project, I ended up putting their `TopDom_v0.0.*.R` scripts into a proper R package (this **TopDom** package) so I could validate the[ir] code via `R CMD check` and so on. Then I started to add a bit of help documentation to make my own life easier.  I also found a few bugs that I fixed and did some improvements but I really tried not to diverge from the original functionality.

Now, since I find it a waste of resources if someone else has to go through the same efforts that I had to, I decided to make this package public.  The goal is also to submit the **TopDom** package to CRAN so that the TopDom method is properly archived for reproducible purposes.  In order to do this, the original authors agreed on releasing their original TopDom scripts under the GPL license, which is also the license of the **TopDom** package.

Having said this, please note that I don't intend to do user support for the TopDom method, how to use it, add new features, and so on.  This is simply because I don't have the resources to do that.  More importantly, I am not the best person to answer questions on the TopDom method and its implementation.  Instead, I recommend that you reach out to the original TopDom authors to get your questions answered.


## Archaeology

The original R script (versions 0.0.1 and 0.0.2) and the TopDom manual were made available on a website of the TopDom author (http<span/>://zhoulab.usc.edu/TopDom/).  As of 2019, that website is no longer available.  Instead, TopDom is now listed as a paragraph on <https://labs.dgsom.ucla.edu/zhou/pages/software> with a link to one of the author GitHub repository [jasminezhoulab/TopDom](https://github.com/jasminezhoulab/TopDom), which holds the TopDom manual and version 0.0.2 of the script but not version 0.0.1.



[R]: https://www.r-project.org/
[TopDom]: https://github.com/HenrikBengtsson/TopDom/
[TopDomData]: https://github.com/HenrikBengtsson/TopDomData/

## Installation
R package TopDom is available on [CRAN](https://cran.r-project.org/package=TopDom) and can be installed in R as:
```r
install.packages("TopDom")
```


### Pre-release version

To install the pre-release version that is available in Git branch `develop` on GitHub, use:
```r
remotes::install_github("HenrikBengtsson/TopDom", ref="develop")
```
This will install the package from source.  

<!-- pkgdown-drop-below -->



NEWS
Package: TopDom
===============

Version: 0.10.1-9001 [2021-11-12]

DOCUMENTATION:

 * Fix equation format issue and update one URL.
 

Version: 0.10.1 [2020-05-04]

CRAN POLICIES:

 * Update URL that otherwise redirects HTTP to HTTPS.
 

Version: 0.10.0 [2020-05-03]

NEW FEATURES:

 * Package now includes mouse Chr19 data from the TopDom study. They can
   be found in the system.file("exdata", package = "TopDom") folder.

 * The orignal TopDom scripts 'TopDom_v0.0.1.R' and 'TopDom_v0.0.2.R' are
   now distributed part of the package as-is.  They can be found in the
   system.file("original-scripts", package = "TopDom") folder.

 * readHiC() gained arguments '...' which is passed as-is to read.table().
 

Version: 0.9.1 [2020-04-02]

BUG FIXES:

 * ggCountHeatmap() for TopDomData could produce a warning on a partial
   argument name.


Version: 0.9.0 [2020-02-19]

SIGNIFICANT CHANGES:

 * The list data.frame elements returned by overlapScores() now has column
   'chromosome' as the first position.  The data.frame:s are of kind tibble.
 
NEW FEATURES:

 * Add as_tibble() for TopDomOverlapScores.
 
DOCUMENTATION:

 * Add further documentation on the 'window.size' parameter.

 * Add reference to Hanjun Shin's PhD thesis.
 

Version: 0.8.2 [2019-12-17]

DOCUMENTATION:

 * Improved help on overlapScores() and TopDom().

 * Provide a reference for the default value for `window.size` of TopDom().
 

Version: 0.8.1 [2019-06-19]

SOFTWARE QUALITY:

 * Fix two cases of partial argument name.


Version: 0.8.0 [2019-04-11]

NEW FEATURES:

 * The TopDom object returned by TopDom() now has an attribute 'parameters'
   which records the value of arguments 'window.size' and 'statFilter'.

 * Made TopDom() faster and more memory efficient by lower the number of
   replicated computations.
 

Version: 0.7.1 [2019-04-09]

ROBUSTNESS:

 * Add internal sanity checks to TopDom() asserting that the intermediate and
   final results are of proper length and does not contain missing values.
 
BUG FIXES:

 * Internal Convert.Bin.To.Domain.TMP() used by TopDom() could produce 'Error
   in `[<-.data.frame`(`*tmp*`, , "to.coord", value = c(NA, 2500, 247500 : 
   replacement has 3 rows, data has 1' because it assumed at least one domain
   was identified.


Version: 0.7.0 [2019-03-15]

SIGNIFICANT CHANGES:

 * Renamed fields returned by overlapScores() to be in singular form, e.g.
   'best_score' instead of 'best_scores'.

NEW FEATURES:

 * Now overlapScores() returns also the lengths of the reference domains.
 

Version: 0.6.0 [2019-01-08]

SIGNIFICANT CHANGES:

 * Renamed and swapped the order of the first two arguments of overlapScores()
   and renamed the second argument to 'reference'.  This was done in order to
   make it clear which set of topological domains the overlap scores are
   calculated relative to.
 

Version: 0.5.0 [2018-11-12]

SIGNIFICANT CHANGES:

 * Lead TopDom author Xianghong Jasmine Zhou confirms by email that the
   original TopDom scripts, and thereby this package, may be released under
   the GNU Public License (GPL).
 

Version: 0.4.0 [2018-10-23]

NEW FEATURES:

 * Add countsPerRegion() for calculating the total contact-frequency counts
   per region specified, e.g. per domain.

 * Add print(), dim(), `[()` and subsetByRegion() for TopDom objects where
   the number of rows in the dimension reflect the number of TopDom domains.
 

Version: 0.3.0 [2018-10-08]

NEW FEATURES:

 * The legacy TopDom() functions, available via legacy(), also accept
   'TopDomData' objects as returned by readHiC().  This is supported mostly
   to make it possible to efficiently compare the different implementations.

 * Added `[`() for TopDomData objects, e.g. tdd[1:100].
 
 * Added subsetByRegion() for TopDomData objects.

 * Added ggCountHeatmap(), ggDomain(), and ggDomainLabel() for TopDomData objects.


Version: 0.2.0 [2018-07-26]

SIGNIFICANT CHANGES:

 * See 'BUG FIXES' below.
 
NEW FEATURES:

 * Add function legacy() for access to the original TopDom v0.0.1 and
   TopDom v0.0.2 implementations, e.g. TopDom::legacy("0.0.1")$TopDom().

BUG FIXES:

 * All TopDom functions except TopDom() itself were the ones from
   TopDom v0.0.2.


Version: 0.1.2 [2018-06-28]

SIGNIFICANT CHANGES:

 * Released on GitHub.
 
NOTE:

 * The license of the underlying TopDom R code/scripts is still unknown,
   i.e. to be decided by the original authors of TopDom.  Any mentioning
   of code licenses in this package / git repository history is invalid.
   
NEW FEATURES:

 * Add overlapScores().

 * Add image() for TopDomData.
 
 * List returned by TopDom() gained class 'TopDom'.

 * Added logical option 'TopDom.debug', which controls whether functions
   produce debug output or not.  The default is FALSE.

DOCUMENTATION:

 * Updated the 'Value' section of help("TopDom") with details from the
   TopDom Manual (an online PDF) provided by Shin et al.
 

Version: 0.1.1 [2018-05-04]

NEW FEATURES:

 * Add print() method for TopDomData object.

 * Reference the TopDom paper (Shin et al., 2016) in the help and the README.
 

Version: 0.1.0 [2018-04-23]

SIGNIFICANT CHANGES:

 * Turned the original TopDom R script into a package.

 * All progress messages are outputted done to standard error.

 * Add readHiC().

NEW FEATURES:

 * TopDom() can now read, via readHiC(), a pure count matrix file without bin
   information.  To read such files, specify what chromosome is being read
   (argument 'chr') and the bin size of the count matrix (argument 'binSize').

 * If the matrix file is not of a known format, then TopDom() produces an
   informative error. Previously it gave a message on stdout and returned 0.

CODE STYLE:

 * Tidied up code.

SOFTWARE QUALITY:

 * Added package tests.
 

Version: 0.0.2 [2016-07-08]

 * TopDom v0.0.2 script from http://zhoulab.usc.edu/TopDom/ with the below
   entries from the official release note:
   
 * Gap Identification module is changed.
   

R/overlapScores.R
#' Calculates Overlap Scores Between Two Sets of Topological Domains
#' 
#' @param a,reference Topological domain (TD) set \eqn{A} and TD reference
#' set \eqn{R} both in a format as returned by [TopDom()].
#'
#' @param debug If `TRUE`, debug output is produced.
#' 
#' @return
#' Returns a named list of class `TopDomOverlapScores`, where the names
#' correspond to the chromosomes in domain reference set \eqn{R}.
#' Each of these chromosome elements contains a data.frame with fields:
#'
#' * `chromosome`  - \eqn{D_{R,c}} character strings
#' * `best_score`  - \eqn{D_{R,c}} numerics in \eqn{[0,1]}
#' * `best_length` - \eqn{D_{R,c}} positive integers
#' * `best_set`    - list of \eqn{D_{R,c}} index vectors
#'
#' where \eqn{D_{R,c}} is the number of TDs in reference set \eqn{R} on
#' chromosome \eqn{c}.  If a TD in reference \eqn{R} is not a `"domain"`,
#' then the corresponding `best_score` and `best_length` values are
#' `NA_real_` and `NA_integer_`, respectively, while `best_set` is an empty
#' list.
#'
#' @details
#' The _overlap score_, \eqn{overlap(A', r_i)}, represents how well a
#' _consecutive_ subset \eqn{A'} of topological domains (TDs) in \eqn{A}
#' overlap with topological domain \eqn{r_i} in reference set \eqn{R}.
#' For each reference TD \eqn{r_i}, the _best match_ \eqn{A'_{max}} is
#' identified, that is, the \eqn{A'} subset that maximize
#' \eqn{overlap(A', r_i)}.
#' For exact definitions, see Page 8 in Shin et al. (2016).
#'
#' Note that the overlap score is an asymmetric score, which means that
#' `overlapScores(a, b) != overlapScores(b, a)`.
#'
#' @section Warning - This might differ not be the correct implementation:
#' The original TopDom scripts do not provide an implementation for
#' calculating overlap scores.  Instead, the implementation of
#' `TopDom::overlapScores()` is based on the textual description of
#' overlap scores provided in Shin et al. (2016).  It is not known if this
#' is the exact same algorithm and implementation as the authors of the
#' TopDom article used.
#'
#' @example incl/overlapScores.R
#' 
#' @references
#' * Shin et al.,
#'   TopDom: an efficient and deterministic method for identifying
#'   topological domains in genomes,
#'   _Nucleic Acids Research_, 44(7): e70, April 2016.
#'   doi: 10.1093/nar/gkv1505,
#'   PMCID: [PMC4838359](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4838359/),
#'   PMID: [26704975](https://pubmed.ncbi.nlm.nih.gov/26704975/)
#'
#' @author Henrik Bengtsson - based on the description in Shin et al. (2016).
#' 
#' @seealso [TopDom].
#'
#' @importFrom tibble as_tibble tibble
#' @export
overlapScores <- function(a, reference, debug = getOption("TopDom.debug", FALSE)) {
  stopifnot(inherits(reference, "TopDom"), inherits(a, "TopDom"))
  stopifnot(is.logical(debug), length(debug) == 1L, !is.na(debug))

  ## Extract the 'domain' fields from the 'reference' and the 'a' TopDom objects
  domains_R <- reference$domain
  domains_A <- a$domain
  
  chromosomes <- unique(domains_R$chr)
  scores <- vector("list", length = length(chromosomes))
  names(scores) <- chromosomes
  for (chr in chromosomes) {
    if (debug) message(sprintf("Chromosome %s ...", chr))
    doms_R <- domains_R[domains_R$chr == chr, ]
    doms_A <- domains_A[domains_A$chr == chr, ]
    scores_chr <- overlapScoresOneChromosome(doms_A, doms_R = doms_R, debug = debug)
    scores_chr <- as_tibble(cbind(tibble(chromosome = chr), scores_chr))
    scores[[chr]] <- scores_chr
    if (debug) message(sprintf("Chromosome %s ... done", chr))
  }

  class(scores) <- c("TopDomOverlapScores", class(scores))
  scores
}


#' @importFrom tibble as_tibble
overlapScoresOneChromosome <- function(doms_A, doms_R, debug = getOption("TopDom.debug", FALSE)) {
  stopifnot(is.logical(debug), length(debug) == 1L, !is.na(debug))
  ## FIXME: Assert that A and reference R are sorted by (chr, pos)

  dtags <- diff(c(0L, as.integer(doms_A$tag == "domain"), 0L))
  sets <- data.frame(
    first = which(dtags == +1L),
    last  = which(dtags == -1L) - 1L,
    stringsAsFactors = FALSE
  )
  sets$from.coord <- doms_A$from.coord[sets$first]
  sets$to.coord <- doms_A$to.coord[sets$last]

  doms_R$length <- as.integer(doms_R$to.coord - doms_R$from.coord)
  doms_A$length <- as.integer(doms_A$to.coord - doms_A$from.coord)

  best_scores <- rep(NA_real_, times = nrow(doms_R))
  best_lengths <- rep(NA_integer_, times = nrow(doms_R))
  best_sets <- vector("list", length = nrow(doms_R))
  idxs_td <- which(doms_R$tag == "domain")
  for (ii in seq_along(idxs_td)) {
    idx_td <- idxs_td[ii]
    if (debug) message(sprintf("TD \"domain\" #%d of %d ...", ii, length(idxs_td)))
    td_R <- doms_R[idx_td, ]
    
    ## Identify sets to consider
    ## Q. How many sets can match this? [0,1], [0,2], [0,3], or even more?
    sets_t <- sets[(sets$to.coord >= td_R$from.coord &
                    sets$from.coord <= td_R$to.coord), ]

    best_score <- 0.0
    best_set <- integer(0L)
    
    ## For each possible A' set ...
    for (kk in seq_len(nrow(sets_t))) {
      set <- sets_t[kk,]
      doms <- doms_A[set$first:set$last,]
      doms$cap <- doms$length
      
      ## TD in A' that is overlapping part of the beginning of reference R 
      before <- which(doms$from.coord <= td_R$from.coord)
      if (length(before) > 0L) {
        before <- before[length(before)]
        doms$cap[before] <- doms$to.coord[before] - td_R$from.coord
      }
      
      ## TD in A' that is overlapping part of the end of reference R
      after <- which(doms$to.coord >= td_R$to.coord)
      if (length(after) > 0L) {
        after <- after[1L]
        doms$cap[after] <- td_R$to.coord - doms$from.coord[after]
      }
      
      ## TDs in A' that are strictly overlapping with reference R
      is_inside <- (doms$from.coord >= td_R$from.coord &
                    doms$to.coord <= td_R$to.coord)
      idxs_t <- c(before, which(is_inside), after)
      n <- length(idxs_t)
      stop_if_not(n > 0L)
      caps <- doms$cap[idxs_t]
      stop_if_not(length(caps) > 0L, any(is.finite(caps)))
      stop_if_not(!anyNA(caps))
      cups <- doms$length[idxs_t]
      stop_if_not(length(cups) > 0L, any(is.finite(cups)))
      if (n == 1L) {
        idxs_u <- list(1L)
        max_score <- caps / cups
      } else if (n == 2L) {
        idxs_u <- list(1L, 1:2, 2L)
      } else if (n >= 3L) {
        idxs_u <- list(1L, 1L:(n-1L), 2L:(n-1L), 2L:n, n)
      }
      stop_if_not(!anyNA(idxs_u))
      scores <- sapply(idxs_u, FUN = function(idxs) {
        sum(caps[idxs]) / sum(cups[idxs])
      })
      stop_if_not(any(is.finite(scores)))
      max_idx <- which.max(scores)

      max_score <- scores[max_idx]
      if (max_score > best_score) {
        best_score <- max_score
        best_set <- idxs_u[[max_idx]]
      }
    } ## for (kk ...)

    ## Sanity checks
    stop_if_not(length(best_score) == 1L, length(td_R$length) == 1L)

    best_scores[ii] <- best_score
    best_lengths[ii] <- td_R$length
    best_sets[[ii]] <- best_set
    
    if (debug) message(sprintf("TD \"domain\" #%d of %d ... done", ii, length(idxs_td)))
  } ## for (ii ...)

  ## Sanity checks
  stop_if_not(length(best_scores) == nrow(doms_R),
              length(best_lengths) == nrow(doms_R),
              length(best_sets) == nrow(doms_R))

  res <- data.frame(best_score = best_scores, best_length = best_lengths, stringsAsFactors = FALSE)
  res$best_set <- best_sets

  res
} ## overlapScoresOneChromosome()


#' @importFrom tibble as_tibble
#' @export
as_tibble.TopDomOverlapScores <- function(x, ...) {
  do.call(rbind, args = x)
}


#' @export
print.TopDomOverlapScores <- function(x, ...) {
  cat(sprintf("%s:\n", paste(class(x), collapse = ", ")))

  cat(sprintf("Chromosomes: [n = %d] %s\n",
              length(x), paste(sQuote(names(x)), collapse = ", ")))

  lengths <- lapply(x, FUN = `[[`, "best_length")
  lengths[["whole genome"]] <- unlist(lengths, use.names = FALSE)
  cat("Summary of reference domain lengths:\n")
  t <- t(sapply(lengths, FUN = function(x) {
    c(summary(x), count = length(x))
  }))
  print(t)
  
  scores <- lapply(x, FUN = `[[`, "best_score")
  scores[["whole genome"]] <- unlist(scores, use.names = FALSE)
  cat("Summary of best scores:\n")
  t <- t(sapply(scores, FUN = function(x) {
    c(summary(x), count = length(x))
  }))
  
  print(t)
}
